testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,297 @@
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_RDONLY);
+	tc.cpsize = SAFE_READ(0, tc.cpfd, &tc.cpattern, 1048);
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+
+	if (tc.cpsize == 1048)
+		tst_brk(TCONF, "Current core pattern is too large to save");
+
+	tc.cpattern[tc.cpsize + 1] = '\0';
+	tst_res(TINFO, "System core pattern is '%s'", tc.cpattern);
+
+	tc.cpfd = SAFE_OPEN(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	SAFE_WRITE(1, tc.cpfd, tmpcpattern, strlen(tmpcpattern));
+	SAFE_CLOSE(tc.cpfd);
+	tc.cpfd = 0;
+}
+
+static void restore_cpattern(void)
+{
+	int fd, rval;
+	size_t written;
+
+	if (tc.cpfd != 0 && close(tc.cpfd))
+		tst_res(TINFO | TERRNO, "close(tc.cpdf)");
+
+	if (tc.cpsize < 1)
+		return;
+
+	fd = open(CORE_PATTERN, O_TRUNC | O_WRONLY);
+	if (fd == -1) {
+		tst_res(TINFO | TERRNO, "open(CORE_PATTERN, O_TRUNC)");
+		goto error;
+	}
+
+	written = write(fd, tc.cpattern, tc.cpsize);
+	if (written < (size_t)tc.cpsize) {
+		tst_brk(TINFO | TERRNO, "write(fd, tc.pattern, ...)");
+		goto error;
+	}
+
+	rval = close(fd);
+	if (rval == -1) {
+		tst_brk(TINFO | TERRNO, "close(fd)");
+		goto error;
+	}
+
+	tc.cpsize = 0;
+	return;
+error:
+	tst_brk(TBROK,
+		"COULD NOT RESTORE "CORE_PATTERN"! It used to be '%s'",
+		tc.cpattern);
+}
+
+static void cleanup(void)
+{
+	restore_cpattern();
+
+	if (tc.fmem != NULL && munmap(tc.fmem, tc.fmemsize))
+		tst_brk(TBROK | TERRNO, "munmap(tc.fmem)");
+	tc.fmem = NULL;
+
+	if (tc.fd > 0 && close(tc.fd))
+		tst_brk(TBROK | TERRNO, "close(tc.fd)");
+
+	if (tc.dfd > 0 && close(tc.dfd))
+		tst_brk(TBROK | TERRNO, "close(tc.dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char c, expectc = 'x';
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TCONF | TERRNO,
+			"Dump file was not found; assuming this is because "
+			"core dumps are disabled and not a bug");
+	tc.dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	while (SAFE_READ(0, tc.dfd, &c, 1)) {
+		switch (c) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == tc.ycount)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+	}
+	return 0;
+}
+
+static void run_child(int advice)
+{
+	char *advstr =
+		advice == MADV_DONTDUMP ? "MADV_DONTDUMP" : "MADV_DODUMP";
+
+	if (madvise(tc.fmem, tc.fmemsize, advice) == -1) {
+		tst_res(TFAIL | TERRNO,
+			"madvise(%p, %lu, %s) = -1",
+			tc.fmem,
+			tc.fmemsize,
+			advstr);
+		return;
+	}
+	abort();
+}
+
+static void run(void)
+{
+	int i, status;
+	pid_t pid;
+	char *fmemc;
+
+	tc.fmem = SAFE_MMAP(NULL,
+			    tc.fmemsize,
+			    PROT_READ | PROT_WRITE,
+			    MAP_ANONYMOUS | MAP_PRIVATE,
+			    -1,
+			    0);
```

Comment:
Here as well, the ```run()``` function can be called in a loop so this ```mmap()``` may leak memory.

Why don't we do the ```mmap()``` in child as well?